### PR TITLE
feat(rt): add support for HTTP proxies

### LINKED ISSUE #494: HTTP Proxy Support
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue, please leave a comment

## Describe the Feature

Support proxy configuration:
* directly as part of `HttpClientEngineConfig`
* auto discovery from JVM settings?

## Additional Context

- [ ] I may be able to implement this feature request

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

closes #494 

## Description of changes

Adds support for common HTTP proxy setup including JVM system properties and environment variables

Does not include support for:
* https proxies (that is proxies that use TLS, you can still use HTTPS origins by tunneling through an HTTP proxy via `CONNECT` request). 
* socks proxy
* any authentication other than `Basic`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/protocol/http-client-engines/test-suite/jvm/src/aws/smithy/kotlin/runtime/http/test/util/KtorOkHttpEngine.kt [hunk 1]
```diff
@@ -43,3 +45,8 @@ internal fun KtorOkHttpEngine(config: HttpClientEngineConfig = HttpClientEngineC
 
     return KtorEngine(okHttpEngine)
 }
+internal fun KtorOkHttpEngine(block: HttpClientEngineConfig.Builder.() -> Unit): HttpClientEngine {
```
Comment (ianbotsf) [inline runtime/protocol/http-client-engines/test-suite/jvm/src/aws/smithy/kotlin/runtime/http/test/util/KtorOkHttpEngine.kt:48]: **Nit**: Missing empty line

File: runtime/protocol/http-client-engines/test-suite/jvm/test/aws/smithy/kotlin/runtime/http/test/ProxyTest.kt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    @Test
+    fun testHttpProxy() = testEngines(
+        // we would expect a customer to configure proxy support on the underlying engine
+        skipEngines = setOf("KtorEngine")
+    ) {
+        engineConfig {
+            val proxyPort = mitmProxy.getMappedPort(8080)
+            proxySelector = ProxySelector { _ ->
+                ProxyConfig.Http("http://127.0.0.1:$proxyPort")
+            }
```
Comment (ianbotsf) [inline runtime/protocol/http-client-engines/test-suite/jvm/test/aws/smithy/kotlin/runtime/http/test/ProxyTest.kt:63]: **Nit**: Unnecessary `_ ->`

File: runtime/protocol/http-client-engines/test-suite/jvm/test/aws/smithy/kotlin/runtime/http/test/ProxyTest.kt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            val call = client.call(req)
+            try {
+                // will be a 301 for http -> https if proxy isn't setup or not configured properly
+                assertEquals(HttpStatusCode.OK, call.response.status, "${client.engine}")
+                val body = call.response.body.readAll()!!.decodeToString()
+                assertEquals("hello proxy", body)
+            } finally {
+                call.complete()
+            }
+        }
```
Comment (ianbotsf) [inline runtime/protocol/http-client-engines/test-suite/jvm/test/aws/smithy/kotlin/runtime/http/test/ProxyTest.kt:126]: **Question**: I see a lot of duplication between this test and the one above it. Can they be refactored to use more common code?

File: runtime/protocol/http-client-engines/test-suite/proxy-scripts/fakeupstream.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+"""
```
Comment (ianbotsf) [inline runtime/protocol/http-client-engines/test-suite/proxy-scripts/fakeupstream.py:1]: **Comment**: This file probably needs a copyright header just like our Kotlin files do.

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt [hunk 5]
```diff
@@ -112,3 +119,60 @@ internal fun CoroutineContext.derivedName(name: String): CoroutineName {
...
+}
+
+private val emptyProxyList = mutableListOf<Proxy>()
+
+internal class OkHttpProxySelector(
+    private val sdkSelector: SdkProxySelector
+) : ProxySelector() {
+    override fun select(uri: URI?): MutableList<Proxy> {
+        if (uri == null) return emptyProxyList
+        val url = uri.toUrl()
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt:164]: It looks like compiler is happy with `List` or `MutableList` and intellij suggested `MutableList` when scaffolding the implementation and I didn't check the assumption. I will update it to use an immutable list

File: runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt [hunk 6]
```diff
@@ -112,3 +119,60 @@ internal fun CoroutineContext.derivedName(name: String): CoroutineName {
...
+
+        return when (val proxyConfig = sdkSelector.select(url)) {
+            is ProxyConfig.Http -> {
+                val okProxy = Proxy(Proxy.Type.HTTP, InetSocketAddress(proxyConfig.url.host, proxyConfig.url.port))
+                return mutableListOf(okProxy)
+            }
+            else -> emptyProxyList
+        }
+    }
+    override fun connectFailed(uri: URI?, sa: SocketAddress?, ioe: IOException?) {
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt:174]: This seems like something that should be enforced by ktlint/format if we are going to go that route...

**Review by ianbotsf — APPROVED** [review]

**Pull request merged**